compiler/damlc/tests/src/DA/Test/DamlcUpgrades.hs

```diff
@@ -0,0 +1,525 @@
+-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Test.DamlcUpgrades (main) where
+
+{- HLINT ignore "locateRunfiles/package_app" -}
+
+import Control.Monad.Extra
+import DA.Bazel.Runfiles
+import Data.Foldable
+import System.Directory.Extra
+import System.FilePath
+import System.IO.Extra
+import DA.Test.Process
+import Test.Tasty
+import Test.Tasty.HUnit
+import SdkVersion
+import DA.Daml.LF.Ast.Version
+import Text.Regex.TDFA
+import qualified Data.Text as T
+import Data.Maybe (fromMaybe)
+
+main :: IO ()
+main = do
+    damlc <- locateRunfiles (mainWorkspace </> "compiler" </> "damlc" </> exe "damlc")
+    defaultMain $ tests damlc
+
+tests :: FilePath -> TestTree
+tests damlc =
+    testGroup
+        "Upgrade"
+        [ test
+              "Fails when template changes signatories"
+              (Just "Message: \n\ESC\\[0;91merror type checking template MyLib.A signatories:\n  The upgraded template A cannot change the definition of its signatories.")
+              [ ( "daml/MyLib.daml"
+                , unlines
+                      [ "module MyLib where"
+                      , "template A with"
+                      , "    p : Party"
+                      , "    q : Party"
+                      , "  where signatory [p]"
+                      ]
+                )
+              ]
+              [ ("daml/MyLib.daml"
+                , unlines
+                      [ "module MyLib where"
+                      , "template A with"
+                      , "    p : Party"
+                      , "    q : Party"
+                      , "  where signatory [p, q]"
+                      ]
+                )
+              ]
+        , test
+              "Fails when template changes observers"
+              (Just "Message: \n\ESC\\[0;91merror type checking template MyLib.A observers:\n  The upgraded template A cannot change the definition of its observers.")
+              [ ( "daml/MyLib.daml"
+                , unlines
+                      [ "module MyLib where"
+                      , "template A with"
+                      , "    p : Party"
+                      , "    q : Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "    observer p"
+                      ]
+                )
+              ]
+              [ ("daml/MyLib.daml"
+                , unlines
+                      [ "module MyLib where"
+                      , "template A with"
+                      , "    p : Party"
+                      , "    q : Party"
+                      , "  where"
+                      , "    signatory p"
+                      , "    observer p, q"
+                      ]
+                )
+              ]
+        , test
+              "Fails when template changes ensure"
```

Comment:
This should succeed. See https://docs.google.com/document/d/18XKYP-PvrhgaDRPnG30H2JZvBawqz5nBpa7gIFsYOwE/edit#bookmark=id.ytdhmxd08boq

Reply:
(The idea is to be able to use "ensure false" as a way to be able to disable templates.)

Reply:
The typechecking doesn't have a warning mechanism and adding it is proving large - splitting into own PR.